core/sr-primitives/src/sr_arithmetic.rs

```diff
@@ -0,0 +1,1097 @@
+				assert_eq!($name::from_percent(2).square(), $name::from_parts(4 * $max / 10_000));
+			}
+
+			#[test]
+			fn per_things_div_works() {
+				// normal
+				assert_eq!($name::from_percent(10) / $name::from_percent(20),
+					$name::from_percent(50)
+				);
+				assert_eq!($name::from_percent(10) / $name::from_percent(10),
+					$name::from_percent(100)
+				);
+				assert_eq!($name::from_percent(10) / $name::from_percent(0),
+					$name::from_percent(100)
+				);
+
+				// will not overflow
+				assert_eq!($name::from_percent(10) / $name::from_percent(5),
+					$name::from_percent(100)
+				);
+				assert_eq!($name::from_percent(100) / $name::from_percent(50),
+					$name::from_percent(100)
+				);
+			}
+		}
+	};
+}
+
+implement_per_thing!(
+	Permill,
+	test_permill,
+	[u32, u64, u128],
+	1_000_000u32,
+	u32,
+	u64,
+	"_Parts per Million_",
+);
+implement_per_thing!(
+	Perbill,
+	test_perbill,
+	[u32, u64, u128],
+	1_000_000_000u32,
+	u32,
+	u64,
+	"_Parts per Billion_",
+);
+implement_per_thing!(
+	Perquintill,
+	test_perquintill,
+	[u64, u128],
+	1_000_000_000_000_000_000u64,
+	u64,
+	u128,
+	"_Parts per Quintillion_",
+);
+implement_per_thing!(
+	Percent,
+	test_per_cent,
+	[u32, u64, u128],
+	100u32,
+	u32,
+	u32,
+	"_Percent_",
+);
+
+/// An unsigned fixed point number. Can hold any value in the range [-9_223_372_036, 9_223_372_036]
+/// with fixed point accuracy of one billion.
+#[cfg_attr(feature = "std", derive(Debug))]
+#[derive(Encode, Decode, Default, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Fixed64(i64);
+
+/// The accuracy of the `Fixed64` type.
+const DIV: i64 = 1_000_000_000;
+
+impl Fixed64 {
+	/// creates self from a natural number.
+	///
+	/// Note that this might be lossy.
+	pub fn from_natural(int: i64) -> Self {
+		Self(int.saturating_mul(DIV))
+	}
+
+	/// Return the accuracy of the type. Given that this function returns the value `X`, it means
+	/// that an instance composed of `X` parts (`Fixed64::from_parts(X)`) is equal to `1`.
+	pub fn accuracy() -> i64 {
+		DIV
+	}
+
+	/// Raw constructor. Equal to `parts / 1_000_000_000`.
+	pub fn from_parts(parts: i64) -> Self {
+		Self(parts)
+	}
+
+	/// creates self from a rational number. Equal to `n/d`.
+	///
+	/// Note that this might be lossy.
+	pub fn from_rational(n: i64, d: u64) -> Self {
+		Self(
+			((n as i128).saturating_mul(DIV as i128) / (d as i128).max(1))
+			.try_into()
+			.unwrap_or(Bounded::max_value())
+		)
+	}
+
+	/// Performs a saturated multiply and accumulate by unsigned number.
+	///
+	/// Returns a saturated `int + (self * int)`.
+	pub fn saturated_multiply_accumulate<N>(&self, int: N) -> N
+	where
+		N: TryFrom<u64> + From<u32> + UniqueSaturatedInto<u32> + Bounded + Clone + Saturating +
+		ops::Rem<N, Output=N> + ops::Div<N, Output=N> + ops::Mul<N, Output=N> +
+		ops::Add<N, Output=N>,
+	{
+		let parts = self.0;
+		let positive = parts > 0;
+
+		// will always fit.
+		let natural_parts: u64 = (parts / DIV) as u64;
```

Comment:
if parts is negative and inferior to -DIV then this result in a big integer number.
as we compute int + int*self. self being a small less than -1 or a very big less than -1 number doesn't change the result but maybe we should add a comment or restructure the logic so that negative case is handle more understandably.

Reply:
also (but this is an issue in whole substrate maybe) the traits bounded to N doesn't force N to be a positive arithmetic type. In the case N is i128 this function is wrong.

Reply:
I see. Honestly, I entirely missed this aspect and was only thinking about positive numbers when writing this. The `self` can, of course, be negative. 

But at least in this, a fix is simple and when we save `let positive = parts > 0;` we basically kill the sign of the `self` and can just use the `abs()` of `parts` which will fit into a u64 when divided by DIV.

I agree with the general comment about substrate as well, having some `Trait Unsigned` would be good but I am not sure how to make it in a generic way.

Reply:
about Unsigned trait we can do the same as num-traits, they have just https://docs.rs/num-traits/0.2.8/num_traits/sign/trait.Unsigned.html which is an empty trait

Reply:
I went as far as adding that (something almost the same) to `traits` but do we actually want that? In practice, someone can have a `struct CustomType` and implement all the requirements of `SimpleArithmetic` for an and use it in the balances module. Doing `pub trait Unsigned: Num {}` would actually block this and only allow these type to be used for balance etc.